src/NHibernate/Event/Default/DefaultInitializeCollectionEventListener.cs

```diff
@@ -80,24 +80,20 @@ public virtual void OnInitializeCollection(InitializeCollectionEvent @event)
 			}
 
 			var batchSize = persister.GetBatchSize();
-			if (batchSize > 1 && persister.Cache.PreferMultipleGet())
+			CollectionEntry[] collectionEntries = null;
+			var collectionBatch = source.PersistenceContext.BatchFetchQueue.QueryCacheQueue
+			                            ?.GetCollectionBatch(persister, collectionKey, out collectionEntries);
+			if ((collectionBatch != null || batchSize > 1) && persister.Cache.PreferMultipleGet())
```

Comment:
I should have think about it sooner, but the same applies to `PreferMultipleGet`. See its xml doc remarks:
```c#
/// When <see langword="true" />, NHibernate will attempt to get other non initialized proxies or
/// collections from the cache instead of only getting the proxy or collection which initialization
/// is asked for. If this cache implementation does not benefit from batching together get operations,
/// this may result in a performance loss.
/// </para>
/// <para>
/// When <see langword="false" />, NHibernate will still call <see cref="GetMany" /> when it has many
/// gets to perform. Its <see cref="CacheBase" /> default implementation is adequate for this case.
/// </para>
```

So the condition should be:
```suggestion
			if (collectionBatch != null || batchSize > 1 && persister.Cache.PreferMultipleGet())
```